ls1intum/Artemis
# [BLOCKED] Feature/Programming-Exercise: Add support for Apollon diagrams

### Checklist

### Motivation and Context



Atm the programming-exercises instructions only support plantUMLs for rendering Uml diagrams.
However as we have the @ls1intum/apollon library, we should integrate it.

Advantages:
- Much easier to handle through the drag&drop editor
- Reuse of existing diagrams
- Less performance heavy, as we can render the diagram in the client

### Description


Two domain commands will be provided:
- Create/Update apollon diagram
- Insert existing apollon diagram
The apollon diagrams that are inserted into the editor will be saved into our database, but linked to a course (otherwise they could be edited by everyone using artemis).

This PR will not make it possible to replace the plantUML diagrams with apollon diagrams, as we can't atm color specific elements of the uml (this is something we use in the interactive instructions for programming exercises).

- [x] Implement command to insert/edit apollon commands in the markdown editor
- [x] Implement modal for inserting/editing apollon diagrams
- [x] Implement preview in programming exercise instructions (remarkable)
- [x] Implement preview in showdown extension (how to create an extension that generates an async REST call?)
- [ ] Fix issues from Apollon update
- [ ] Add exercise id to apollon diagram model
- [ ] Implement command for inserting existing apollon diagrams of course

### Steps for Testing


1. Log in to ArTEMiS
2. Create/Edit a programming-exercise
3. Insert an apollon diagram by clicking the domain command
4. Check the preview
5. Edit the created apollon diagram by placing the cursor within it and then clicking the domain command 

### Screenshots



**New domain command:**

![apollon_2](https://user-images.githubusercontent.com/28230611/58889340-09b6fd00-86e9-11e9-8c5d-a1820a56e301.png)

**Create/Edit diagram in modal:**

![apollon_3](https://user-images.githubusercontent.com/28230611/58889357-1176a180-86e9-11e9-8a66-8fbcf880f98b.png)

**Representation of diagram in markdown:**

![apollon_4](https://user-images.githubusercontent.com/28230611/58889370-150a2880-86e9-11e9-8ec3-dd5c8a7e21cb.png)

**Rendering of the diagram:**

![apollon_1](https://user-images.githubusercontent.com/28230611/58889381-176c8280-86e9-11e9-8f2b-f23a5a324ede.png)
